core/src/main/scala/kafka/log/remote/RemoteLogManager.scala

```diff
@@ -498,6 +498,112 @@ class RemoteLogManager(fetchLog: TopicPartition => Option[Log],
+
+      The below is a situation where the remote segment has the same leader epoch as the currently
+      ongoing local leader epoch.
+      CORRECT
+      Remote:                                i1 ===================== i2
+      Local :        <=== j1 ===============================================>
+
+      INCORRECT
+      Remote:                                       i1 ===================== i2
+      Local :        <=== j1 ======================================== j2 ===>
+
+      INCORRECT
+      Remote:   i1 ===================== i2
+      Local :        <=== j1 ======================================== j2 ===>
+      */
+      def doesRemoteSegmentHaveGoodLineage(validEpochs: util.TreeMap[Integer, lang.Long],
+                                           segmentMetadata: RemoteLogSegmentMetadata,
+                                           logEndOffset: lang.Long): Boolean = {
+        val segmentEpochs = segmentMetadata.segmentLeaderEpochs()
+        val segmentEndOffset = segmentMetadata.endOffset()
+        val firstSegmentEpochEntry = segmentEpochs.firstEntry()
+        segmentEpochs.entrySet().stream()
+          .allMatch(entry => {
+            val segmentEpochKey = entry.getKey
+            val segmentEpochStartOffset = entry.getValue
+            if (validEpochs.containsKey(segmentEpochKey)) {
+              val validEpochStartOffset = validEpochs.get(segmentEpochKey)
+              val validEpochEndEntry = validEpochs.higherEntry(segmentEpochKey)
+              val remoteEndEntry = segmentEpochs.higherEntry(segmentEpochKey)
+
+              if (validEpochEndEntry == null && remoteEndEntry != null) {
+                // Remote has more entries than local
+                // This should be impossible but we are erring on the side of defining the behaviour
+                val remoteLogSegmentId = segmentMetadata.remoteLogSegmentId()
+                val topicIdPartition = segmentMetadata.topicIdPartition()
+                val extraneousSegmentLeaderEpoch = remoteEndEntry.getKey
+                warn(s"[Segment $remoteLogSegmentId; TopicPartition $topicIdPartition]: Segment has at least one more leader epoch ($extraneousSegmentLeaderEpoch) than locally ($segmentEpochKey)")
+                false
+              } else if (validEpochEndEntry != null && remoteEndEntry == null) {
+                // Remote has fewer entries than local
+                segmentEpochStartOffset >= validEpochStartOffset && segmentEndOffset < validEpochEndEntry.getValue
+              } else if (validEpochEndEntry != null && remoteEndEntry != null) {
+                // This is not the active leader epoch for both remote and local
+                if (firstSegmentEpochEntry.getKey == segmentEpochKey) {
+                  remoteEndEntry.getKey == validEpochEndEntry.getKey &&
+                    segmentEpochStartOffset >= validEpochStartOffset &&
+                    remoteEndEntry.getValue <= validEpochEndEntry.getValue
```

Comment:
Hi Kamal

Consider a case when we have let's say two segments and the epoch ends in the middle of segment 2. In such a case, the validEpochEndEntry will know the exact point of the end for that epoch but segment 1 will assume that the end is >= to it's own end entry (since the end/start entry of a segment is limited to within the segment). Hence, it is very much possible that a valid lineage contains a situation where `remoteEndEntry.getValue <= validEpochEndEntry.getValue` and not strictly equal. 

Does this make sense to you?

Reply:
I still don't get this. This check infers that there can be two different start-offset for a given epoch. In the case mentioned above, `remoteEndEntry` would be null for segment-1, and for segment-2  it should contain the exact point of the end of the previous epoch.

Reply:
When the next epoch's are the same, their startOffsets should also be similar. 

For example: 
A remote segment with offsets [100, 950] has its internal epochs as:
leader epoch vs startOffset is:
1, 100
2, 250

and the next segment with offsets[951, 2380] has its internal segment epoch vs start-offsets as:
2, 951
3, 1790

and leader replica's leader-epoch vs start-offset as:
1, 50
2, 250 
3, 1790

remote segments and leader replica's epoch:start-offsets have the aligned lineage. segment-1 will have the same startoffset for leader epoch 2. 

So, the check should be remoteEndEntry.getValue == validEpochEndEntry.getValue instead of remoteEndEntry.getValue <= validEpochEndEntry.getValue as this epoch entry is not the first epoch entry of the segment but the next epoch which can be with in the segment.

Reply:
Yes, you and @kamalcph are right. Will fix this.

Reply:
This is fixed now in the latest commit.